src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxInputNode.cs

```diff
@@ -61,35 +67,42 @@ public void SaveStateAndFree(ImmutableSegmentedDictionary<object, IStateTable>.B
                 tables[_owner] = _transformTable.ToImmutableAndFree();
             }
 
-            public void VisitTree(Lazy<SyntaxNode> root, EntryState state, SemanticModel? model, CancellationToken cancellationToken)
+            public void VisitTree(Lazy<SyntaxNode> root, EntryState state, int syntaxTreeIndex, IncrementalGeneratorRunStep? inputStep, GeneratorRunStateTable.Builder runStateTableBuilder, SemanticModel? model, CancellationToken cancellationToken)
             {
                 if (state == EntryState.Removed)
                 {
                     // mark both syntax *and* transform nodes removed
-                    _filterTable.RemoveEntries();
-                    _transformTable.RemoveEntries();
+                    _filterTable.TryRemoveEntries(TimeSpan.Zero, ImmutableArray<(IncrementalGeneratorRunStep, int)>.Empty, out ImmutableArray<SyntaxNode> removedNodes);
+
+                    for (int i = 0; i < removedNodes.Length; i++)
+                    {
+                        _transformTable.TryRemoveEntries(TimeSpan.Zero, ImmutableArray<(IncrementalGeneratorRunStep, int)>.Empty);
+                    }
+
                 }
                 else
                 {
                     Debug.Assert(model is object);
 
                     // get the syntax nodes from cache, or a syntax walk using the filter
-                    ImmutableArray<SyntaxNode> nodes;
-                    if (state != EntryState.Cached || !_filterTable.TryUseCachedEntries(out nodes))
+                    if (state != EntryState.Cached || !_filterTable.TryUseCachedEntries(TimeSpan.Zero, ImmutableArray<(IncrementalGeneratorRunStep, int)>.Empty, out ImmutableArray<SyntaxNode> nodes))
                     {
+                        var stopwatch = SharedStopwatch.StartNew();
                         nodes = IncrementalGeneratorSyntaxWalker.GetFilteredNodes(root.Value, _owner._filterFunc, cancellationToken);
-                        _filterTable.AddEntries(nodes, EntryState.Added);
+                        _filterTable.AddEntries(nodes, state, stopwatch.Elapsed, ImmutableArray<(IncrementalGeneratorRunStep, int)>.Empty, state);
                     }
 
                     // now, using the obtained syntax nodes, run the transform
-                    foreach (var node in nodes)
+                    for (int i = 0; i < nodes.Length; i++)
```

Comment:
>for (int i = 0; i < nodes.Length; i++) [](http://example.com/codeflow?start=20&length=38)

Is the change from `foreach` to `for` necessary?

Reply:
It was at one point during this PR. It doesn't look like it is required with the current design. I'll change it back.